### DISCUSSION TITLE
How to preserve newlines and indents with ViewRecord

### DISCUSSION THREAD

**Discussion by hrvoj3e** [filamentphp/filament#4335] [Help]
I would like to use placeholder to show text field from DB but preserve new lines and indents.
Is this maybe built-in or must I implement a view for that? 

My solution ATM is using `<pre>` to preserve newlines and indents.
```php
Forms\Components\ViewField::make('body2')
        ->view('filament.partials.multiline')
        ->viewData(['content' => $record->body])
```
with view
```blade
<pre style="font-family: inherit">
{{ $content }}
</pre>
```
Using `{!! nl2br(e($content)) !!}` will preserve only newlines.

What I expected is 
```php
Forms\Components\Placeholder::make('body')
->content(fn (): string => $record->body)

// this would be great
//   ->multiline()
//  or
//   ->preserveFormatting()
```

Textarea works OK but all other fields on ViewRecord are Placholders so ...
```php
Forms\Components\Textarea::make('body')
```

**Comment by danharrin** [comment]
Probably use HTML? `->content(fn ($record): HtmlString => new HtmlString(Str::markdown($record->body)))`

**Reply by hrvoj3e** [reply]
Tried it just now. It does not preserve indents, and multiple newlines are not preserved.
Only `<pre>` solution works fully.

**Reply by danharrin** [reply]
You could `str_replace` a newline with `<br />`

**Reply by hrvoj3e** [reply]
This will not cover indents (spaces/tabs).

**Comment by hrvoj3e** [comment]
I duplicated placeholder view and extended Placeholder class: `php artisan make:form-field PlaceholderText`
 - I need this in multiple ViewRecord Pages so duplicating code in all those place inside callback seems not a way to go.
 - I need `label()` and using view does not provide this

```php
<?php

namespace App\Filament\Forms\Components;

use Filament\Forms\Components\Placeholder;

class PlaceholderText extends Placeholder
{
    protected string $view = 'filament.forms.components.placeholder-text';
}

```

```blade
<x-dynamic-component
    :component="$getFieldWrapperView()"
    :id="$getId()"
    :label="$getLabel()"
    :label-sr-only="$isLabelHidden()"
    :helper-text="$getHelperText()"
    :hint="$getHint()"
    :hint-icon="$getHintIcon()"
    :state-path="$getStatePath()"
>
    <div {{ $attributes->merge($getExtraAttributes())->class(['filament-forms-placeholder-component']) }}>
    <pre style="font-family: inherit">{{ $getContent() }}</pre>
    </div>
</x-dynamic-component>
```

Is something wrong with this approach?
Could I somehow extend placeholder view `forms::components.placeholder` inside my component view?

**Reply by hrvoj3e** [reply]
Found a shorter way without using additional view.

Would like to know if later there would be a problem with this approach?

```php
namespace App\Filament\Forms\Components;

use Filament\Forms\Components\Placeholder;
use Illuminate\Support\HtmlString;

class PlaceholderText extends Placeholder
{
    public function getContent()
    {
        $content = parent::getContent();

        $content = new HtmlString(sprintf('<pre style="font-family: inherit">%s</pre>', $content));

        return $content;
    }
}
```
